rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+be configured to store events temporarily in a disk buffer. Since the
+events are not actually delivered when they are buffered, the final
+delivery status will have to be confirmed after the event is loaded back
+from the buffer. The source, to be used to determine where
+to provide the finalization status indicator, must be converted into a form
+that may be serialized (ie string or integer index). However, the
+configuration may be reloaded while the event is buffered in such a way
+that the source name provided in the configuration may change, or a
+different component may be substituted for the same identifier. As such,
+the configured component name is not sufficient to uniquely identify the
+source. Further, Vector may be stopped while the event is buffered and
+then restarted with an identical configuration. At this point, the old
+source will be distinct from the current source, despite having the same
+configured name. As such, the persisted identifier for the source must
+stay the same across configuration reloads, but must be different for
+each run of Vector.
+
+## Internal Proposal
+
+There are two components to the proposal: the communication between
+event finalization and the sources, and the resulting event
+metadata required to track the event status.
+
+### Event finalization
+
+When instantiated, the topology will provide each source with a unique
+identifier for the token. The source may use this identifier to create
+one or more notification channels, each with their own unique
+identifier. This provides both a unique identifier of the source that
+needs to receive the event finalization status as well as the mechanism
+for delivering the event identifier and status.
+
+The finalization status is recorded when the last copy of the event is
+dropped. At that time, the finalization status stored in the event is
+recorded in a batch status buffer. Since events may be internally
+dropped when they are merged, a special "no-op" status is used to
+indicate that no status needs to be recorded to prevent superfluous
+status updates. When the last copy of the batch status is dropped, it is
+sent to the source.
+
+### Event metadata
+
+Given the channel above, the following source metadata will be added to events:
+
+1. A finalization status, to be delivered to all sources.
+2. An array of event sources.
+
+Each event source is a shared reference to a batch notifier, containing:
+
+1. The final status indicator. This will default to a success indicator,
+   but may be set to indicate failure if any event in the batch fails
+   delivery.
+2. A one-shot notification channel to the source.
+3. The unique identifier used to recreate the channel after deserialization.
+
+### Data structures
+
+```rust
+struct EventMetadata {
+    // … existing fields …
+    finalizer: Option<Arc<EventFinalizer>>,
+}
+
+struct EventFinalizer {
+    status: EventStatus,
+    sources: Box<[Arc<BatchNotifier>]>,
```

Comment:
The struct is that each event has an optional finalizer, and each finalizer references one or more source batches. When an event is fanned out, it is cloned to each destination, but each destination needs a reference to the same finalizer, hence the outer `Arc<EventFinalizer>`. Similarly, all events in a batch need to reference the same batch, hence the inner `Arc<BatchNotifier>`. I wish I could avoid that, but it seems necessary.

Given that this will already be heap allocated, there probably isn't much difference between the boxed slice and `Vec<Arc<BatchNotifier>>`. My thinking was that event merging would be relatively rare, and would produce a new finalizer anyways, which would create a new sources array.

Reply:
I buy it.